packages/nextjs/src/server/keyless-node.ts

```diff
@@ -0,0 +1,177 @@
+  }
+  const gitignorePath = nodeRuntime.path.join(process.cwd(), '.gitignore');
+  if (!existsSync(gitignorePath)) {
+    writeFileSync(gitignorePath, '');
+  }
+
+  // Check if `.clerk/` entry exists in .gitignore
+  const gitignoreContent = readFileSync(gitignorePath, 'utf-8');
+  if (!gitignoreContent.includes(CLERK_HIDDEN + '/')) {
+    appendFileSync(gitignorePath, `\n${CLERK_HIDDEN}/\n`);
+  }
+}
+
+const getKeylessConfigurationPath = () => {
+  if (!nodeRuntime.path) {
+    throw "Clerk: fsModule.path is missing. This is an internal error. Please contact Clerk's support.";
+  }
+  return nodeRuntime.path.join(process.cwd(), CLERK_HIDDEN, '.tmp', 'keyless.json');
+};
+
+const getKeylessReadMePath = () => {
+  if (!nodeRuntime.path) {
+    throw "Clerk: fsModule.path is missing. This is an internal error. Please contact Clerk's support.";
+  }
+  return nodeRuntime.path.join(process.cwd(), CLERK_HIDDEN, '.tmp', 'README.md');
+};
+
+let isCreatingFile = false;
+
+function safeParseClerkFile(): AccountlessApplication | undefined {
+  if (!nodeRuntime.fs) {
+    throw "Clerk: fsModule.fs is missing. This is an internal error. Please contact Clerk's support.";
+  }
+  const { readFileSync } = nodeRuntime.fs;
+  try {
+    const CONFIG_PATH = getKeylessConfigurationPath();
+    let fileAsString;
+    try {
+      fileAsString = readFileSync(CONFIG_PATH, { encoding: 'utf-8' }) || '{}';
+    } catch {
+      fileAsString = '{}';
+    }
+    return JSON.parse(fileAsString) as AccountlessApplication;
+  } catch {
+    return undefined;
+  }
+}
+
+const createMessage = (keys: AccountlessApplication) => {
+  return `\n\x1b[35m\n[Clerk]:\x1b[0m You are running on keyless mode.\nYou can \x1b[35mclaim your keys\x1b[0m by visiting ${keys.claimUrl}\n`;
+};
+
+async function createOrReadKeyless(): Promise<AccountlessApplication | undefined> {
+  if (!nodeRuntime.fs) {
+    // This should never happen.
+    throw "Clerk: fsModule.fs is missing. This is an internal error. Please contact Clerk's support.";
+  }
+  const { existsSync, writeFileSync, mkdirSync, rmSync } = nodeRuntime.fs;
+
+  /**
+   * If another request is already in the process of acquiring keys return early.
+   * Using both an in-memory and file system lock seems to be the most effective solution.
+   */
+  if (isCreatingFile || existsSync(CLERK_LOCK)) {
+    return undefined;
+  }
+
+  isCreatingFile = true;
+
+  writeFileSync(
+    CLERK_LOCK,
+    // In the rare case, the file persists give the developer enough context.
+    'This file can be deleted. Please delete this file and refresh your application',
+    {
+      encoding: 'utf8',
+      mode: '0777',
+      flag: 'w',
+    },
+  );
+
+  const CONFIG_PATH = getKeylessConfigurationPath();
+  const README_PATH = getKeylessReadMePath();
+
+  mkdirSync(nodeRuntime.path ? nodeRuntime.path.dirname(CONFIG_PATH) : '', { recursive: true });
+  mkdirSync(nodeRuntime.path ? nodeRuntime.path.dirname(README_PATH) : '', { recursive: true });
```

Comment:
nit, these are in the same dir:
```suggestion
  mkdirSync(nodeRuntime.path ? nodeRuntime.path.dirname(CONFIG_PATH) : '', { recursive: true });
```